File: implementations/csharp/Provekit.ProofEnvelope/Cbor.cs
// SPDX-License-Identifier: Apache-2.0
//
// Deterministic CBOR encoder. RFC 8949 §4.2.1 "Core Deterministic
// Encoding" rules:
//
//   - shortest-form integer head encoding (smallest of immediate /
//     u8 / u16 / u32 / u64)
//   - definite-length items only
//   - map keys sorted by bytewise lex order of their CBOR-encoded form
//
// We emit only the major types we need: unsigned int, byte string,
// text string, array, map. Mirrors
// implementations/rust/provekit-proof-envelope/src/cbor.rs and
// implementations/cpp/provekit/proof-envelope/cbor.cpp 1:1.

namespace Provekit.ProofEnvelope;

public enum CborMajor : byte
{
    UnsignedInt = 0,
    ByteString = 2,
    TextString = 3,
    Array = 4,
    Map = 5,
}

public static class Cbor
{
    /// <summary>
    /// Append a CBOR head: 1 byte (immediate), 2 bytes (uint8), 3 bytes
    /// (uint16), 5 bytes (uint32), or 9 bytes (uint64). Always shortest
    /// form per §4.2.1.
    /// </summary>
    public static void AppendHead(List<byte> output, CborMajor major, ulong arg)
    {
        var mt = (byte)((byte)major << 5);
        if (arg < 24)
        {
            output.Add((byte)(mt | (byte)arg));
            return;
        }
        if (arg <= 0xFF)
        {
            output.Add((byte)(mt | 24));
            output.Add((byte)arg);
            return;
        }
        if (arg <= 0xFFFF)
        {
            output.Add((byte)(mt | 25));
            output.Add((byte)(arg >> 8));
            output.Add((byte)arg);
            return;
        }
        if (arg <= 0xFFFFFFFF)
        {
            output.Add((byte)(mt | 26));
            output.Add((byte)(arg >> 24));
            output.Add((byte)(arg >> 16));
            output.Add((byte)(arg >> 8));
            output.Add((byte)arg);
            return;
        }
        output.Add((byte)(mt | 27));
        for (var i = 7; i >= 0; i--)
        {
            output.Add((byte)(arg >> (i * 8)));
        }
    }

    public static void EncodeUint(List<byte> output, ulong value) =>
        AppendHead(output, CborMajor.UnsignedInt, value);

    public static void EncodeBstr(List<byte> output, ReadOnlySpan<byte> bytes)
    {
        AppendHead(output, CborMajor.ByteString, (ulong)bytes.Length);
        for (var i = 0; i < bytes.Length; i++) output.Add(bytes[i]);
    }

    public static void EncodeTstr(List<byte> output, string utf8)
    {
        var bytes = System.Text.Encoding.UTF8.GetBytes(utf8);
        AppendHead(output, CborMajor.TextString, (ulong)bytes.Length);
        output.AddRange(bytes);
    }

    public static void EncodeArrayHead(List<byte> output, ulong count) =>
        AppendHead(output, CborMajor.Array, count);

    public static void EncodeMapHead(List<byte> output, ulong count) =>
        AppendHead(output, CborMajor.Map, count);
}

File: implementations/zig/provekit-proof-envelope-zig/src/cbor.zig
// SPDX-License-Identifier: Apache-2.0
//
// Deterministic CBOR encoder. RFC 8949 §4.2.1 rules:
//   - shortest-form integer encoding (smallest of short / u8 / u16 / u32 / u64)
//   - definite-length items only
//   - map keys sorted in bytewise lex order of their CBOR-encoded form
//   - we emit only the major types we need: unsigned int, byte string,
//     text string, array, map.
//
// Mirrors implementations/rust/provekit-proof-envelope/src/cbor.rs 1:1
// and implementations/cpp/provekit/proof-envelope/cbor.cpp.

const std = @import("std");

pub const CborMajor = enum(u8) {
    unsigned_int = 0,
    byte_string = 2,
    text_string = 3,
    array = 4,
    map = 5,
};

pub fn appendHead(out: *std.ArrayList(u8), alloc: std.mem.Allocator, major: CborMajor, arg: u64) !void {
    const mt: u8 = @as(u8, @intFromEnum(major)) << 5;
    if (arg < 24) {
        try out.append(alloc, mt | @as(u8, @intCast(arg)));
        return;
    }
    if (arg <= 0xFF) {
        try out.append(alloc, mt | 24);
        try out.append(alloc, @as(u8, @intCast(arg)));
        return;
    }
    if (arg <= 0xFFFF) {
        try out.append(alloc, mt | 25);
        try out.append(alloc, @as(u8, @intCast(arg >> 8)));
        try out.append(alloc, @as(u8, @intCast(arg & 0xFF)));
        return;
    }
    if (arg <= 0xFFFF_FFFF) {
        try out.append(alloc, mt | 26);
        try out.append(alloc, @as(u8, @intCast(arg >> 24)));
        try out.append(alloc, @as(u8, @intCast((arg >> 16) & 0xFF)));
        try out.append(alloc, @as(u8, @intCast((arg >> 8) & 0xFF)));
        try out.append(alloc, @as(u8, @intCast(arg & 0xFF)));
        return;
    }
    try out.append(alloc, mt | 27);
    var i: u6 = 8;
    while (i > 0) {
        i -= 1;
        try out.append(alloc, @as(u8, @intCast((arg >> (@as(u6, i) * 8)) & 0xFF)));
    }
}

pub fn encodeUint(out: *std.ArrayList(u8), alloc: std.mem.Allocator, value: u64) !void {
    try appendHead(out, alloc, .unsigned_int, value);
}

pub fn encodeBstr(out: *std.ArrayList(u8), alloc: std.mem.Allocator, bytes: []const u8) !void {
    try appendHead(out, alloc, .byte_string, @as(u64, bytes.len));
    try out.appendSlice(alloc, bytes);
}

pub fn encodeTstr(out: *std.ArrayList(u8), alloc: std.mem.Allocator, utf8: []const u8) !void {
    try appendHead(out, alloc, .text_string, @as(u64, utf8.len));
    try out.appendSlice(alloc, utf8);
}

pub fn encodeArrayHead(out: *std.ArrayList(u8), alloc: std.mem.Allocator, count: u64) !void {
    try appendHead(out, alloc, .array, count);
}

pub fn encodeMapHead(out: *std.ArrayList(u8), alloc: std.mem.Allocator, count: u64) !void {
    try appendHead(out, alloc, .map, count);
}

// ---------------------------------------------------------------------------
// Tests: pinned vectors from RFC 8949 §3.3 + the rust kit.
// ---------------------------------------------------------------------------

const testing = std.testing;

test "shortest-form uint: 0..23 single byte" {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(testing.allocator);

    try encodeUint(&buf, testing.allocator, 0);
    try testing.expectEqualSlices(u8, &.{0x00}, buf.items);

    buf.clearRetainingCapacity();
    try encodeUint(&buf, testing.allocator, 23);
    try testing.expectEqualSlices(u8, &.{0x17}, buf.items);
}

test "shortest-form uint: 24..255 two bytes" {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(testing.allocator);

    try encodeUint(&buf, testing.allocator, 24);
    try testing.expectEqualSlices(u8, &.{ 0x18, 24 }, buf.items);

    buf.clearRetainingCapacity();
    try encodeUint(&buf, testing.allocator, 255);
    try testing.expectEqualSlices(u8, &.{ 0x18, 0xff }, buf.items);
}

test "shortest-form uint: 256..65535 three bytes" {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(testing.allocator);

    try encodeUint(&buf, testing.allocator, 256);
    try testing.expectEqualSlices(u8, &.{ 0x19, 0x01, 0x00 }, buf.items);
}

test "shortest-form uint: 65536..2^32-1 five bytes" {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(testing.allocator);

    try encodeUint(&buf, testing.allocator, 65536);
    try testing.expectEqualSlices(u8, &.{ 0x1a, 0x00, 0x01, 0x00, 0x00 }, buf.items);
}

test "tstr emits major-3 head then utf8 bytes" {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(testing.allocator);

    try encodeTstr(&buf, testing.allocator, "hello");
    // major 3 (text), len 5 short form: 0x60 | 5 = 0x65, then "hello"
    try testing.expectEqualSlices(u8, &.{ 0x65, 'h', 'e', 'l', 'l', 'o' }, buf.items);
}

test "bstr emits major-2 head then raw bytes" {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(testing.allocator);

    try encodeBstr(&buf, testing.allocator, &.{ 0xde, 0xad, 0xbe, 0xef });
    try testing.expectEqualSlices(u8, &.{ 0x44, 0xde, 0xad, 0xbe, 0xef }, buf.items);
}

test "map head with 7 entries" {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(testing.allocator);

    try encodeMapHead(&buf, testing.allocator, 7);
    // major 5 (map), count 7 short form: 0xa0 | 7 = 0xa7
    try testing.expectEqualSlices(u8, &.{0xa7}, buf.items);
}
